### PR TITLE
ENH: use same value counts for roll_var to remove floating point artifacts

### LINKED ISSUE #42064: BUG: rolling std can't handle mixture of (relatively) big and small numbers
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

```python
import pandas as pd
import numpy as np

df = pd.DataFrame(np.zeros(1000))
df.iloc[0] = 1000
df.rolling(10).std()

Out[124]: 
            0
0         NaN
1         NaN
2         NaN
3         NaN
4         NaN
..        ...
995  0.000004
996  0.000004
997  0.000004
998  0.000004
999  0.000004
[1000 rows x 1 columns]

```

#### Problem description

If we have a relatively big number at the top, and the remaining is zeros, then the rolling result is not zero

#### Expected Output
```python
df.iloc[0] = 0
df.rolling(10).std()

Out[125]: 
       0
0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
..   ...
995  0.0
996  0.0
997  0.0
998  0.0
999  0.0
[1000 rows x 1 columns]

```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2cb96529396d93b46abab7bbc73a208e708c642e
python           : 3.8.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.8.0-53-generic
Version          : #60-Ubuntu SMP Thu May 6 07:46:32 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 1.2.4
numpy            : 1.20.2
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 52.0.0.post20210125
Cython           : None
pytest           : 6.2.3
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.2
html5lib         : None
pymysql          : 0.9.3
psycopg2         : 2.8.6 (dt dec pq3 ext lo64)
jinja2           : 3.0.0
IPython          : 7.23.1
pandas_datareader: 0.9.0
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.4
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 4.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.6.2
sqlalchemy       : 1.4.15
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.53.1

</details>

### PR DESCRIPTION
**Pull request by auderson**
ENH: use same value counts for roll_var to remove floating point artifacts
Please refer to https://github.com/pandas-dev/pandas/issues/42064#issuecomment-1072934337 for the explaination of this new method.
- [ ] closes #42064 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by auderson** [comment]
Looks like there's a network issue in the test.

File: pandas/_libs/window/aggregations.pyx [hunk 1]
```diff
@@ -307,6 +308,14 @@ cdef inline void add_var(float64_t val, float64_t *nobs, float64_t *mean_x,
         return
 
     nobs[0] = nobs[0] + 1
+
+    # GH#42064, record num of same values to remove floating point artifacts
+    if val == prev_value[0] and val != MAXfloat64 and val != MINfloat64:
```
Comment (jreback) [inline pandas/_libs/window/aggregations.pyx:313]: why do max/min float matter here?

File: pandas/_libs/window/aggregations.pyx [hunk 2]
```diff
@@ -307,6 +308,14 @@ cdef inline void add_var(float64_t val, float64_t *nobs, float64_t *mean_x,
...
+        num_consecutive_same_value[0] += 1
+    else:
+        num_consecutive_same_value[0] = 1  # reset to 1 (include current value itself)
```
Comment (jreback) [inline pandas/_libs/window/aggregations.pyx:316]: dont' use line comments, put it above

**Review by jreback — CHANGES_REQUESTED** [review]
we need a battery of tests that exercise this code.

you can use something like the doc examples as tests. they should fail before and pass after.

need to exercise np.inf and np.nan as well (as you are including this in the code)'

pls parameterize over both std and var

**Comment by mroeschke** [comment]
Isn't there logic that needs to be added to `remove_var` too? Essentially if we need to track if each window has all unique, non-nan values, then this sliding window algorithm needs to account for values moving out of the window?

File: pandas/_libs/window/aggregations.pyx (hunk 1 above)
Comment (auderson) [reply pandas/_libs/window/aggregations.pyx:313]: It seems infs will be converted to nans in `_prep_values` already: https://github.com/pandas-dev/pandas/blob/afec0e9c1b2856cf28071327141a0ac0e4ba8aee/pandas/core/window/rolling.py#L354-L377
Then this inf check will be unnecessary

**Comment by auderson** [comment]
> Isn't there logic that needs to be added to `remove_var` too? Essentially if we need to track if each window has all unique, non-nan values, then this sliding window algorithm needs to account for values moving out of the window?

Here's my thoughts:

Suppose we have x number of same values, then the window enters into some nans. Now `same_value_count` is still x. Then we enter a window with valid numbers.

If the new number is still the same, `same_value_count` will continue to increment by 1. Though it's actutally larger than num of same values in current window, we're safe to set result as zero. If the new number differs,  `same_value_count` will be reset to 1.

So in summary, I think it's OK that `same_value_count` is not updated in `remove_var`, because `same_value_count` basically records the num of same value in the entire series.

**Comment by auderson** [comment]
I will write some tests in the upcoming days, but I'm not sure where should I put them. Do I need to add some in `pandas/tests/window/test_rolling.py`?

If only doc test is needed, do I just have to write some examples in rolling.var, std 's docs (in fact I already update the old examples in the docs)?

**Comment by mroeschke** [comment]
Gotcha. Not sure if there's a condition where `remove_var` is called in an iteration while `add_var` is not which could also maybe mess up the counting(?), but if not then the resetting should work. 

`pandas/tests/window/test_rolling.py` is fine for the tests. I don't think we specifically need more doc examples for this.

It would be good to have an example like

`data = pd.Series([0, 0, 0, 1, 2, 3, 3, 3, 2, 1]).rolling(3).var()`

But the `0, 1, 2, 3` are numbers where the floating point artifacts used to show up

**Comment by auderson** [comment]
> Gotcha. Not sure if there's a condition where `remove_var` is called in an iteration while `add_var` is not which could also maybe mess up the counting(?), but if not then the resetting should work.

The case where only `remove_var` is called happens when it enters a series of NaN or it goes to the end. For that first case, I've added corresponding tests.

Basically, the counting needn't care about removed values, as long as we know how many `x` have appeared when we see it, and the count is correct, then it's OK even if it starts from the beginning of the array. And we only need to make sure every time we see a new number in `add_var`, the count updates correctly.

**Comment by auderson** [comment]
The test failed in `py38_32bit`. Do I need to do anything about it? Seems like a bizarre error

**Comment by auderson** [comment]
Not so familiar with git. After I clicked "sync fork" in pycharm, my branch was added commits from others which may be the cause of these test failures(?) 😓

My last version failed in `py38_32bit` and I can't figure out why. @mroeschke Would you tell me what I should do next?

**Comment by mroeschke** [comment]
I would recommend following the contributing guide to set up git correctly (not sure how Pycharm decided to do this): https://pandas.pydata.org/docs/development/contributing.html#working-with-the-code

Namely you probably need to merge the main branch into your branch

```
git remote add upstream https://github.com/pandas-dev/pandas.git
git checkout roll_var_remove_floating_point_artifacts
git fetch upstream
git merge upstream/main
```

**Comment by auderson** [comment]
@mroeschke I reset my branch to the earliest state and merge with main and then force-pushed it. It really works.
The tests all passed expect for the doc build. But that seems to be a bug in main?

File: pandas/_libs/window/aggregations.pyx [hunk 3]
```diff
@@ -278,15 +278,15 @@ def roll_mean(const float64_t[:] values, ndarray[int64_t] start,
 
 
 cdef inline float64_t calc_var(int64_t minp, int ddof, float64_t nobs,
-                               float64_t ssqdm_x) nogil:
+                               float64_t ssqdm_x, int64_t num_consecutive_same_value) nogil:
```
Comment (mroeschke) [inline pandas/_libs/window/aggregations.pyx:281]: Can probably use `Py_ssize_t` instead of `int64_t`

File: pandas/_libs/window/aggregations.pyx [hunk 4]
```diff
@@ -376,9 +385,13 @@ def roll_var(const float64_t[:] values, ndarray[int64_t] start,
             # never removed
             if i == 0 or not is_monotonic_increasing_bounds or s >= end[i - 1]:
 
+                prev_value = values[s]
+                num_consecutive_same_value = 0
```
Comment (mroeschke) [inline pandas/_libs/window/aggregations.pyx:389]: This can be set in `cdef`

File: pandas/tests/window/test_rolling.py [hunk 5]
```diff
@@ -1781,3 +1781,103 @@ def test_step_not_integer_raises():
...
+    ],
+)
+def test_rolling_var_same_value_count_logic(values, window, min_periods, expected):
+    # GH 42064
+
+    sr = Series(values)
+    result_var = sr.rolling(window, min_periods=min_periods).var()
+    # 1. result should be close to correct value
+    # non-zero values can still differ slightly as the result of online algorithm
+    assert np.isclose(result_var, expected, equal_nan=True).all()
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:1876]: Could you just use `tm.assert_series_equal` since the results should be more "exact" now?

These tests also need to be replicated in `window/test_numba.py`

File: pandas/_libs/window/aggregations.pyx (hunk 4 above)
Comment (auderson) [reply pandas/_libs/window/aggregations.pyx:389]: Should them be reset when `not is_monotonic_increasing_bounds or s >= end[i - 1]`? I don't quite understand when will these conditions be true, so I moved them from cdef to here.

File: pandas/tests/window/test_rolling.py (hunk 5 above)
Comment (auderson) [reply pandas/tests/window/test_rolling.py:1876]: While adding test for numba, I found a bug here:
https://github.com/pandas-dev/pandas/blob/625dc0f082a53674f0ea32b12f629f2eda00f285/pandas/core/_numba/kernels/sum_.py#L84
which should be `if nobs == 0 == min_periods`.

**Comment by auderson** [comment]
@mroeschke looks like `Py_ssize_t` can't pass 32bit test

File: pandas/_libs/window/aggregations.pyx (hunk 4 above)
Comment (mroeschke) [reply pandas/_libs/window/aggregations.pyx:389]: I think we support rolling if the index is monotonically decreasing for example (might be good to add a test for that).

Overall, it seem like `num_consecutive_same_value = 0` will always be set (because `i==0`) so better to add it in the cdef?

Comment (auderson) [reply pandas/_libs/window/aggregations.pyx:389]: So is it a feature that will be supported in the future?

 If that's the case, the count should be __reset__ in every window or it'll be messed up.

 For current use case where `not is_monotonic_increasing_bounds or s >= end[i - 1]` is always not True, we can move it to cdef.

Comment (mroeschke) [reply pandas/_libs/window/aggregations.pyx:389]: I do see we have some tests with monotonically decreasing index so it appears supported

`test_rolling_on_decreasing_index`
`test_rolling_decreasing_indices`
`test_rolling_decreasing_indices_centered`

Comment (auderson) [reply pandas/_libs/window/aggregations.pyx:389]: I've read these tests. Though they're named _decreasing_index_, they are still computed from top to bottom. It's just the __Index__ that is inverse.

For example in `test_rolling_on_decreasing_index`:
https://github.com/pandas-dev/pandas/blob/9a2d9ea214ea65ad93149c4987260254b5fe1329/pandas/tests/window/test_timeseries_window.py#L731-L744
`window_indexer.get_window_bounds` will return:
image
these indices are still mono-increasing.

Then back to our case in `roll_var`, what I'm thinking is that, if we have real use case in the future that we actully compute from the last row up to the first row (mono-decreasing bounds), we should place reset in the if block. Otherwise if we can make sure all rolling algos is from top to bottom, we can place it to cdef.

Anyway, it should be safe to place it to cdef for now. Cuz if there are mono-decr bounds, the `roll_var` algo is false too. It should clearly have to remove obs from the tail and this logic is not present yet.

Comment (auderson) [reply pandas/_libs/window/aggregations.pyx]: `prev_value` can't be set in cdef. If length is 0, it'll cause a segfault.

Comment (auderson) [reply pandas/_libs/window/aggregations.pyx:389]: It's so strange that this version will casuse the same error as I used Py_ssize_t for count type in 32 bit test.
47d5e970832424a85c6d387e22bbdde
6cead89f87de3fe11b311cf44f05fa0
Revert it back now.

Comment (mroeschke) [reply pandas/_libs/window/aggregations.pyx:389]: Ah I see what you mean now. Yes, rolling will never be evaluated "bottom" to "top" though the index may be ordered in a monotonically decreasing fashion.

File: pandas/tests/window/test_rolling.py [hunk 6]
```diff
@@ -1781,3 +1781,105 @@ def test_step_not_integer_raises():
...
+    expected = Series(expected)
+
+    sr = Series(values)
+    result_var = sr.rolling(window, min_periods=min_periods).var()
+    # 1. result should be close to correct value
+    # non-zero values can still differ slightly from "truth"
+    # as the result of online algorithm
+    tm.assert_series_equal(result_var, expected)
+    # 2. zeros should be exactly the same since the new algo takes effect here
+    tm.assert_series_equal(expected == 0, result_var == 0)
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:1880]: This check should be redundant with the above one since `tm.assert_series_equal` compares values exactly.

Comment (auderson) [reply pandas/tests/window/test_rolling.py:1880]: If I set `check_exact=True` (default=False)，it can't pass 32-bit test due to precision loss. So I have to assert those zeros seperately.

Comment (auderson) [reply pandas/tests/window/test_rolling.py:1880]: I'll try using UDF to generate expected data later.

Comment (mroeschke) [reply pandas/tests/window/test_rolling.py:1880]: Ah okay. This is fine then (no need for the UDF). Instead maybe comment that it's needed for the 32 bit build

Comment (auderson) [reply pandas/tests/window/test_rolling.py:1891]: Sure.

Comment (auderson) [reply pandas/tests/window/test_rolling.py:1891]: Looks like there's a bug in `click`

File: pandas/tests/window/test_rolling.py [hunk 7]
```diff
@@ -1781,3 +1781,116 @@ def test_step_not_integer_raises():
...
+    ],
+)
+def test_rolling_var_same_value_count_logic(values, window, min_periods, expected):
+    """
+    For GH 42064.
+
+    With new algo implemented, result will be set to .0 in rolling var
+    if sufficient amount of consecutively same values are found.
+
+    Below, use `assert_series_equal` twice to check for equality,
```
[the remark this answers is not in the document]
Comment (auderson) [reply pandas/tests/window/test_rolling.py]: updated

**Review by mroeschke — COMMENTED** [review]
Looks pretty good. 

Could you also modify `test_rolling_var_floating_artifact_precision` and `test_rolling_std_small_values` and remove the `atol` and `rtol` parameter (since they shouldn't be needed anymore) and add a `# GH <PR number>` in the test?

**Comment by auderson** [comment]
@mroeschke 
I'm thinking if I should update the algo for other rolling functions as well. 
Today I encountered a bug caused by rolling mean:
```python
pd.Series([ -0.022556,  0.046154, -0.051471, -0.015504,
       -0.03937 ,  0.008197, -0.02439 , -0.05    , -0.052632, -0.046296,
        0.      ,  0.      ,  0.      ,  0.      ,  0.      ,  0.      ,
        0.      ,  0.      ]).rolling(5).mean()
```
That outputs:
```
0              NaN
1              NaN
2              NaN
3              NaN
4    -1.654940e-02
5    -1.039880e-02
6    -2.450760e-02
7    -2.421340e-02
8    -3.163900e-02
9    -3.302420e-02
10   -3.466360e-02
11   -2.978560e-02
12   -1.978560e-02
13   -9.259200e-03
14    1.387779e-18
15    1.387779e-18
16    1.387779e-18
17    1.387779e-18
dtype: float64
```
This is really annoying. Previously I thought these floating artifacts would not happen in `mean` as it's implemented with Kahan summation. But now I'd say maybe we need to add same-value counts for `mean` \ `sum`, too?

**Comment by mroeschke** [comment]
>  I'm thinking if I should update the algo for other rolling functions as well.

Sure, I think this would be better in a follow-up PR

**Comment by auderson** [comment]
@mroeschke 
> Sure, I think this would be better in a follow-up PR

Should I open seperate PRs for different functions or merge them in a bigger PR? I'm looking into `sum`, `mean`, `std`, `var`, `skew`, `kurt`. 

As for `skew` & `kurt`, they currently output NaN for small divisors:
https://github.com/pandas-dev/pandas/blob/672798ec7b5565366be24aa4f57f6374f3a5e0fd/pandas/_libs/window/aggregations.pyx#L418-L434

```python
r = pd.Series([ -0.022556,  0.046154, -0.051471, -0.015504,
       -0.03937 ,  0.008197, -0.02439 , -0.05    , -0.052632, -0.046296,
        0.      ,  0.      ,  0.      ,  0.      ,  0.      ,  0.      ,
        0.      ,  0.      ]).rolling(5)
r.skew()
```
```
0          NaN
1          NaN
2          NaN
3          NaN
4     1.482404
5     0.655211
6     0.444657
7     0.516469
8     1.289529
9     1.387878
10    1.182903
11    0.577833
12   -0.639457
13   -2.236068
14         NaN
15         NaN
16         NaN
17         NaN
dtype: float64
```
the last 4 number is NaN instead of 0.
`r.kurt` also outputs NaN.

**Comment by jreback** [comment]
do not merge to anything bigger

wait for review and merge here

**Comment by auderson** [comment]
The former tests here require "all rolling kurt for all equal values should return Nan" (which referenced a wrong issue id?#18804)
With the new impl, the default result of same values is 0 for skew and -3 for kurt.
Possiblly related: #30993

image

**Review by mroeschke — COMMENTED** [review]
This looked pretty good, just needs to resolve the merge conflict

**Comment by auderson** [comment]
> This looked pretty good, just needs to resolve the merge conflict

I put these changes in __Other enhancements__ section, not sure if it's right.

File: doc/source/whatsnew/v1.5.0.rst [hunk 8]
```diff
@@ -95,6 +95,9 @@ Other enhancements
 - :meth:`pd.concat` now raises when ``levels`` is given but ``keys`` is None (:issue:`46653`)
 - :meth:`pd.concat` now raises when ``levels`` contains duplicate values (:issue:`46653`)
 - Added ``numeric_only`` argument to :meth:`DataFrame.corr`, :meth:`DataFrame.corrwith`, and :meth:`DataFrame.cov` (:issue:`46560`)
+- :meth:`DataFrame.rolling.var`, :meth:`DataFrame.rolling.std`, :meth:`Series.rolling.var`, :meth:`Series.rolling.std` now generate correct result 0 for window containing same values (:issue:`42064`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:98]: move these to bug fixes in the rolling bug fix section.

File: pandas/tests/window/test_rolling.py [hunk 9]
```diff
@@ -1203,6 +1203,9 @@ def test_rolling_var_numerical_issues(func, third_value, values):
     result = getattr(ds.rolling(2), func)()
     expected = Series([np.nan] + values)
     tm.assert_series_equal(result, expected)
+    # GH 42064
+    # new `roll_var` will output 0.0 correctly
+    tm.assert_series_equal(result == 0, expected == 0)
```
Comment (jreback) [inline pandas/tests/window/test_rolling.py:1208]: this test doesn't make sense. is the expected wrong?

File: pandas/tests/window/test_rolling.py [hunk 10]
```diff
@@ -1769,3 +1775,113 @@ def test_step_not_integer_raises():
...
+    # use `assert_series_equal` twice to check for equality,
+    # because `check_exact=True` will fail in 32-bit tests due to
+    # precision loss.
+
+    # 1. result should be close to correct value
+    # non-zero values can still differ slightly from "truth"
+    # as the result of online algorithm
+    tm.assert_series_equal(result_var, expected)
+    # 2. zeros should be exactly the same since the new algo takes effect here
+    tm.assert_series_equal(expected == 0, result_var == 0)
```
Comment (jreback) [inline pandas/tests/window/test_rolling.py:1882]: again i am not sure what this is testing, we are *already* testing the result_var is equal to the expected. how does this help?

**Review by jreback — CHANGES_REQUESTED** [review]
can you make an additional test that tests something like [5, 7, 5, 7, 5, 7] e.g. nothing is repeated yet its likey degenerate.

File: pandas/tests/window/test_rolling.py (hunk 9 above)
Comment (mroeschke) [reply pandas/tests/window/test_rolling.py:1208]: Apparently with `check_exact=True` (default is False & checks within a tolerance), these tests fail the 32 bit build.

https://github.com/pandas-dev/pandas/pull/46431#discussion_r836983428=

I think this is okay since it specifically asserts the 0's

Comment (auderson) [reply pandas/tests/window/test_rolling.py:1208]: > I think this is okay since it specifically asserts the 0's
Yeah, though it looks redundant.

File: doc/source/whatsnew/v1.5.0.rst [hunk 11]
```diff
@@ -94,7 +94,7 @@ Other enhancements
 - :meth:`DataFrame.reset_index` now accepts a ``names`` argument which renames the index names (:issue:`6878`)
 - :meth:`pd.concat` now raises when ``levels`` is given but ``keys`` is None (:issue:`46653`)
 - :meth:`pd.concat` now raises when ``levels`` contains duplicate values (:issue:`46653`)
-- Added ``numeric_only`` argument to :meth:`DataFrame.corr`, :meth:`DataFrame.corrwith`, and :meth:`DataFrame.cov` (:issue:`46560`)
+- Added ``numeric_only`` argument to :meth:`DataFrame.corr`, :meth:`DataFrame.corrwith`, and :meth:`DataFrame.cov` (:issue:`46560`)\
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:97]: There's an unneeded `\` at the end here

File: doc/source/whatsnew/v1.5.0.rst [hunk 12]
```diff
@@ -178,10 +178,87 @@ Styler
 
 - Fix showing "None" as ylabel in :meth:`Series.plot` when not setting ylabel (:issue:`46129`)
 
-.. _whatsnew_150.notable_bug_fixes.notable_bug_fix2:
+.. _whatsnew_150.notable_bug_fixes.rolling_moments:
 
-notable_bug_fix2
-^^^^^^^^^^^^^^^^
+rolling mean/sum/std/var/skew/kurt now output correct result for window with same values
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:183]: This can just be a 1 line description in the `Groupby/resample/rolling` section

File: doc/source/whatsnew/v1.5.0.rst (hunk 11 above)
Comment (auderson) [reply doc/source/whatsnew/v1.5.0.rst:97]: oh, I'll remove that

File: doc/source/whatsnew/v1.5.0.rst (hunk 12 above)
Comment (auderson) [reply doc/source/whatsnew/v1.5.0.rst:183]: Did you mean to write like this?

image

Comment (mroeschke) [reply doc/source/whatsnew/v1.5.0.rst:183]: Just a 1 line statement like the rest of the notes in this section 

``` - Bug in ... (:issue:`42064`) ```

Comment (mroeschke) [reply doc/source/whatsnew/v1.5.0.rst:183]: You can use the old line you had before creating this section.

File: doc/source/whatsnew/v1.5.0.rst [hunk 13]
```diff
@@ -180,8 +180,10 @@ Styler
 
 .. _whatsnew_150.notable_bug_fixes.notable_bug_fix2:
 
-notable_bug_fix2
-^^^^^^^^^^^^^^^^
+Groupby/resample/rolling
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+- rolling mean / sum / std / var / skew / kurt now output correct result for window with same values (:issue:`42064`, :issue:`30993`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:186]: There's an _existing_ `Groupby/resample/rolling` below where this whatsnew note can go.

Also this should just mention `var/std` for this PR. Your other PR's can include the other functions

Comment (auderson) [reply doc/source/whatsnew/v1.5.0.rst:186]: My bad, I didn't notice that.

**Review by mroeschke — APPROVED** [review]
LGTM on green

**Comment by mroeschke** [comment]
@auderson could you merge main one more time?

**Review by jreback — APPROVED** [review]
lgtm @mroeschke merge on green

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Awesome thanks @auderson (failure unrelated)!

**Comment by auderson** [comment]
Nice👍
